src/lib/shim/src/lib.rs

```diff
@@ -24,9 +34,191 @@ pub fn manager_shmem() -> Option<&'static ManagerShmem> {
+
+        // Save to thread-local, so that we can deallocate on thread exit
+        assert!(
+            THREAD_SIGNAL_STACK.get().replace(new_stack).is_null(),
+            "Allocated signal stack twice for current thread"
+        );
+
+        // Set up guard page
+        unsafe { rustix::mm::mprotect(new_stack, 4096, rustix::mm::MprotectFlags::empty()) }
+            .unwrap();
+
+        // Install via `sigaltstack`. The kernel will switch to this stack when
+        // invoking one of our signal handlers.
+        let stack_descriptor = linux_api::signal::stack_t {
+            ss_sp: new_stack,
+            ss_size: SHIM_SIGNAL_STACK_SIZE.try_into().unwrap(),
+            // Clear the alternate stack settings on entry to signal handler, and
+            // restore it on exit.  Otherwise a signal handler invoked while another
+            // is running on the same thread would clobber the first handler's stack.
+            // Instead we want the second handler to push a new frame on the alt
+            // stack that's already installed.
+            ss_flags: linux_api::signal::SigAltStackFlags::SS_AUTODISARM.bits(),
+        };
+        unsafe {
+            linux_api::signal::sigaltstack(Some(&stack_descriptor), None).unwrap();
+        }
+    }
+
+    /// # Safety
+    ///
+    /// The current thread must exit before:
+    /// * Another thread runs.
+    /// * That thread also frees its stack.
+    /// * This thread runs again on the signal stack (e.g. by handling a new signal).
+    ///
+    /// Generally in the shim we rely on Shadow's scheduling model to ensure
+    /// this, since we know Shadow won't permit another thread to run
+    /// preemptively before the curent thread has a chance to finish exiting.
+    pub unsafe fn free() {
+        // A signal stack waiting to be freed.
+        //
+        // We can't free the current thread's signal stack, since we may be running on it.
+        // Instead we save the pointer to this global, and free the pointer that was already
+        // there, if any.
+        //
+        // We smuggle the pointer through as a `usize`, since pointers aren't `Sync`.
+        static FREE_SIGNAL_STACK: SelfContainedMutex<usize> = SelfContainedMutex::const_new(0);
+
+        let mut free_signal_stack = FREE_SIGNAL_STACK.lock();
+        let this_thread_stack = THREAD_SIGNAL_STACK.get().replace(core::ptr::null_mut());
+        let stack_to_free_now =
+            core::mem::replace(&mut *free_signal_stack, this_thread_stack as usize);
+        if stack_to_free_now != 0 {
+            unsafe {
+                rustix::mm::munmap(
+                    stack_to_free_now as *mut core::ffi::c_void,
+                    SHIM_SIGNAL_STACK_SIZE,
+                )
+            }
+            .unwrap();
+        }
+    }
+}
+
+mod tls_ipc {
+    use super::*;
+    static IPC_DATA_BLOCK: ShimTlsVar<RefCell<Option<ShMemBlockAlias<IPCData>>>> =
+        ShimTlsVar::new(|| RefCell::new(None));
```

Comment:
Comment that this is used to communicate / pass messages with the main shadow process?